### PR TITLE
NUM-384 filters account by their balance

### PR DESCRIPTION
**Pull request by jdupas22**
This feature allows us to filters the accounts we get using the /{ledger}/accounts route by their balance

The request now takes two more optionnal parameters, <balance> and <balance_operator>

- balance : int64 that represents the balance volume of the account
- balance_operator : comparison operator, can be gt, gte, e, lt, lte

I also tried to make the tests easier to read by bracketing them

### CODE REVIEW

File: pkg/api/controllers/account_controller_test.go [hunk 1]
```diff
@@ -144,6 +144,111 @@ func TestGetAccounts(t *testing.T) {
 				assert.Len(t, resp.Cursor.Data, 1)
 				assert.Equal(t, resp.Cursor.Data[0].Address, "bob")
 
+				// test filter by balance >= 50 with default operator
+				{
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go]: Instead of brackets, I would suggest to use the `t.Run()` method to split in named sub tests, just as I did in this PR: https://github.com/numary/ledger/pull/218

File: pkg/api/controllers/account_controller_test.go [hunk 2]
```diff
@@ -144,6 +144,111 @@ func TestGetAccounts(t *testing.T) {
...
+						"balance":          []string{"100"},
+						"balance_operator": []string{"e"},
+					})
+					assert.Equal(t, http.StatusOK, rsp.Result().StatusCode)
+					resp = getAccountsResponse{}
+					assert.NoError(t, json.Unmarshal(rsp.Body.Bytes(), &resp))
+					assert.Len(t, resp.Cursor.Data, 1)
+					assert.Equal(t, resp.Cursor.Data[0].Address, "bob")
+				}
+
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go:288]: What happens with a `balance` param that cannot be parsed as an integer?

File: pkg/storage/sqlstorage/accounts.go [hunk 3]
```diff
@@ -38,6 +39,34 @@ func (s *Store) accountsQuery(p map[string]interface{}) *sqlbuilder.SelectBuilde
 		}
 	}
 
+	if balance, ok := p["balance"]; ok && balance.(string) != "" {
+
+		sb.Join(s.schema.Table("volumes"), "accounts.address = volumes.account")
+		balanceOperation := "volumes.input - volumes.output "
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go]:
```suggestion
		balanceOperation := "volumes.input - volumes.output"
```

File: pkg/storage/sqlstorage/accounts.go [hunk 4]
```diff
@@ -38,6 +39,34 @@ func (s *Store) accountsQuery(p map[string]interface{}) *sqlbuilder.SelectBuilde
...
+			switch balanceOperator.(string) {
+			case "lte":
+				sb.Where(sb.LessEqualThan(balanceOperation, balanceValue))
+			case "lt":
+				sb.Where(sb.LessThan(balanceOperation, balanceValue))
+			case "gt":
+				sb.Where(sb.GreaterThan(balanceOperation, balanceValue))
+			case "e":
+				sb.Where(sb.Equal(balanceOperation, balanceValue))
+			default: // default to gte
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go]: If the operator is invalid, I think that we should return a bad request error to the user instead of defaulting.

**Review by antoinegelloz — CHANGES_REQUESTED** [review]

File: pkg/api/controllers/account_controller_test.go (hunk 1 above)
Comment (jdupas22) [reply pkg/api/controllers/account_controller_test.go]: you're right, it's better this way, i'll fix it

File: pkg/storage/sqlstorage/accounts.go (hunk 4 above)
Comment (jdupas22) [reply pkg/storage/sqlstorage/accounts.go]: yes, you're right for the operator, i'll fix it

File: pkg/api/controllers/account_controller_test.go (hunk 2 above)
Comment (jdupas22) [reply pkg/api/controllers/account_controller_test.go:288]: i'm going to add an error to the function, and i'll use it for this case and the bad operator

Comment (jdupas22) [reply pkg/api/controllers/account_controller_test.go:288]: after talking with gfyrag, i think the best solution (for now) is to panic in the storage package, but check the parameters in the controller. please tell me what you think

Comment (antoinegelloz) [reply pkg/api/controllers/account_controller_test.go:288]: Yes it's great!

File: pkg/storage/sqlstorage/accounts.go (hunk 4 above)
Comment (antoinegelloz) [reply pkg/storage/sqlstorage/accounts.go]: Perfect :)

File: pkg/api/controllers/account_controller.go [hunk 5]
```diff
@@ -19,6 +20,25 @@ func NewAccountController() AccountController {
...
+	BALANCE_OPERATOR_E   BalanceOperator = "e"
+	BALANCE_OPERATOR_GT  BalanceOperator = "gt"
+	BALANCE_OPERATOR_GTE BalanceOperator = "gte"
+	BALANCE_OPERATOR_LT  BalanceOperator = "lt"
+	BALANCE_OPERATOR_LTE BalanceOperator = "lte"
+)
+
+func (b BalanceOperator) IsValid() bool {
+	switch b {
+	case BALANCE_OPERATOR_E, BALANCE_OPERATOR_GT, BALANCE_OPERATOR_GTE, BALANCE_OPERATOR_LT, BALANCE_OPERATOR_LTE:
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller.go]: I think this line is too long, it would be prettier with one const per line ;)

File: pkg/api/controllers/account_controller.go [hunk 6]
```diff
@@ -38,11 +58,27 @@ func (ctl *AccountController) CountAccounts(c *gin.Context) {
...
+	if balance != "" {
+		if _, err := strconv.ParseInt(balance, 10, 64); err != nil {
+			ResponseError(c, ledger.ValidationError{Msg: "invalid parameter 'balance', should be a number"})
+		}
+	}
+
+	balanceOperator := c.Query("balance_operator")
+	if balanceOperator != "" {
+		if !BalanceOperator(balanceOperator).IsValid() {
+			ResponseError(c, ledger.ValidationError{Msg: "invalid parameter 'balance_operator', should be one of 'e, gt, gte, lt, lte'"})
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller.go]: Line is too long as well!

File: pkg/api/controllers/account_controller_test.go [hunk 7]
```diff
@@ -144,6 +144,120 @@ func TestGetAccounts(t *testing.T) {
...
+					assert.Len(t, resp.Cursor.Data, 1)
+					assert.Equal(t, resp.Cursor.Data[0].Address, "bob")
+				})
+
+				t.Run("invalid balance", func(t *testing.T) {
+					assert.Panics(t, func() {
+						internal.GetAccounts(h, url.Values{
+							"balance": []string{"toto"},
+						})
+					}, "should panic")
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go]: The Store method is supposed to panic but shouldn't we check in this test that a bad request is returned to the user?
`assert.Equal(t, http.StatusBadRequest, rsp.Result().StatusCode)`

File: pkg/api/controllers/account_controller_test.go [hunk 8]
```diff
@@ -144,6 +144,120 @@ func TestGetAccounts(t *testing.T) {
...
+				})
+
+				t.Run("invalid balance_operator", func(t *testing.T) {
+					assert.Panics(t, func() {
+						internal.GetAccounts(h, url.Values{
+							"balance":          []string{"100"},
+							"balance_operator": []string{"toto"},
+						})
+					}, "should panic")
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go]: Same as above :)

File: pkg/storage/sqlstorage/accounts.go [hunk 9]
```diff
@@ -38,6 +39,38 @@ func (s *Store) accountsQuery(p map[string]interface{}) *sqlbuilder.SelectBuilde
...
+		if balanceOperator, ok := p["balance_operator"]; ok && balanceOperator.(string) != "" {
+			switch balanceOperator.(string) {
+			case "lte":
+				sb.Where(sb.LessEqualThan(balanceOperation, balanceValue))
+			case "lt":
+				sb.Where(sb.LessThan(balanceOperation, balanceValue))
+			case "gte":
+				sb.Where(sb.GreaterEqualThan(balanceOperation, balanceValue))
+			case "gt":
+				sb.Where(sb.GreaterThan(balanceOperation, balanceValue))
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts.go:67]: Missing a `greater than` test, thanks Codecov 😅

**Review by antoinegelloz — COMMENTED** [review]
What do you think about also adding unit tests in the sqlstorage package for this new feature?

File: pkg/api/controllers/account_controller_test.go (hunk 7 above)
Comment (jdupas22) [reply pkg/api/controllers/account_controller_test.go]: yeah you're right. I forgot to add a 'return' after the call to ResponseError, so it panicked here too. I fixed it, i'll add the panic test in storage

**Comment by jdupas22** [comment]
@antoinegelloz @gfyrag maybe we try to merge https://github.com/numary/ledger/pull/216 before this one ?

**Comment by antoinegelloz** [comment]
> @antoinegelloz @gfyrag maybe we try to merge #216 before this one ?

Yes we could :) Did you have a look at it already?

**Comment by flemzord** [comment]
@all-contributors please add @jdupas22  for code

File: pkg/api/controllers/account_controller.go [hunk 10]
```diff
@@ -19,6 +20,29 @@ func NewAccountController() AccountController {
 	return AccountController{}
 }
 
+type BalanceOperator string
+
+const (
+	BALANCE_OPERATOR_E   BalanceOperator = "e"
```
Comment (gfyrag) [inline pkg/api/controllers/account_controller.go]: Usually, constants in go are in CamelCase.

Comment (jdupas22) [reply pkg/api/controllers/account_controller.go]: Oh right ! i'm fixing that

File: pkg/api/controllers/account_controller.go [hunk 11]
```diff
@@ -19,6 +20,29 @@ func NewAccountController() AccountController {
 	return AccountController{}
 }
 
+type BalanceOperator string
```
Comment (gfyrag) [inline pkg/api/controllers/account_controller.go]: Maybe move on storage package and use on file sqlstorage/account.go to avoid copy/paste?

File: pkg/api/controllers/account_controller.go [hunk 12]
```diff
@@ -38,11 +62,35 @@ func (ctl *AccountController) CountAccounts(c *gin.Context) {
...
+		}
+	}
+
 	cursor, err := l.(*ledger.Ledger).GetAccounts(
 		c.Request.Context(),
 		query.After(c.Query("after")),
 		query.Address(c.Query("address")),
 		query.Metadata(c.QueryMap("metadata")),
+		query.Balance(balance),
+		query.BalanceOperator(balanceOperator),
```
Comment (gfyrag) [inline pkg/api/controllers/account_controller.go]: Maybe make query.BalanceOperator() accept a BalanceOperator type instead of a string.

**Review by gfyrag — DISMISSED** [review]

File: pkg/api/controllers/script_controller_test.go [hunk 13]
```diff
@@ -103,7 +102,7 @@ func TestPostScriptPreview(t *testing.T) {
 				internal.Decode(t, rsp.Body, &res)
 
 				store := internal.GetStore(t, driver, ctx)
-				cursor, err := store.GetTransactions(ctx, query.New())
+				cursor, err := store.GetTransactions(ctx, storage.New())
```
Comment (gfyrag) [inline pkg/api/controllers/script_controller_test.go]: NewQuery maybe

File: pkg/api/controllers/account_controller.go [hunk 14]
```diff
@@ -38,11 +40,40 @@ func (ctl *AccountController) CountAccounts(c *gin.Context) {
...
+		if balanceOperator, ok = storage.NewBalanceOperator(balanceOperatorStr); !ok {
+			ResponseError(
+				c,
+				ledger.ValidationError{
+					Msg: "invalid parameter 'balance_operator', should be one of 'e, gt, gte, lt, lte'"})
+			return
+		}
+	} else {
+		// default value for balanceOperator
+		balanceOperator = storage.BalanceOperatorGte
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller.go]: Just a minor suggestion :) 
Instead of comment + assignment, a new `const DefaultBalanceOperator = BalanceOperatorGte` in the storage package could be created, and used as the init value line 55.
What do you think?

File: pkg/storage/sqlstorage/accounts_test.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+	assert.NoError(t, err)
+
+	_, err = store.Initialize(context.Background())
+	assert.NoError(t, err)
+
+	q := storage.Query{
+		Params: map[string]interface{}{
+			"balance": "toto",
+		},
+	}
```
Comment (antoinegelloz) [inline pkg/storage/sqlstorage/accounts_test.go]: should be moved into the sub test `panic invalid balance`

File: pkg/storage/utils.go [hunk 16]
```diff
@@ -1,8 +1,6 @@
-package query
+package storage
```
Comment (antoinegelloz) [inline pkg/storage/utils.go]: Why did you rename this file to `utils.go`, I find the previous name `query.go` more appropriate :)

File: pkg/storage/utils.go [hunk 17]
```diff
@@ -16,6 +14,37 @@ type Query struct {
 
 type Modifier func(*Query)
```
Comment (antoinegelloz) [inline pkg/storage/utils.go]: Now that the package name changed from `query` to `storage`, I think that this should be renamed to bring more meaning:
```suggestion
type QueryModifier func(*Query)
```
What do you think?

File: pkg/storage/utils.go [hunk 18]
```diff
@@ -16,6 +14,37 @@ type Query struct {
...
+
+func NewBalanceOperator(s string) (BalanceOperator, bool) {
+	if !BalanceOperator(s).IsValid() {
+		return "", false
+	}
+
+	return BalanceOperator(s), true
+}
+
 func New(qms ...[]Modifier) Query {
```
Comment (antoinegelloz) [inline pkg/storage/utils.go]: Same reason as above:
```suggestion
func NewQuery(qms ...[]Modifier) Query {
```

File: pkg/api/controllers/script_controller_test.go (hunk 13 above)
Comment (antoinegelloz) [inline pkg/api/controllers/script_controller_test.go]: I feel like we are loosing meaning here :/

File: pkg/api/controllers/transaction_controller.go [hunk 19]
```diff
@@ -26,10 +26,10 @@ func (ctl *TransactionController) CountTransactions(c *gin.Context) {
 
 	count, err := l.(*ledger.Ledger).CountTransactions(
 		c.Request.Context(),
-		query.Reference(c.Query("reference")),
-		query.Account(c.Query("account")),
-		query.Source(c.Query("source")),
-		query.Destination(c.Query("destination")),
+		storage.Reference(c.Query("reference")),
```
Comment (antoinegelloz) [inline pkg/api/controllers/transaction_controller.go]: Also here :/

File: pkg/storage/utils.go (hunk 16 above)
Comment (jdupas22) [reply pkg/storage/utils.go]: i put the corresponding objects (transactions or accounts) in their own files, and appended 'Query' to most of the objects/method so that it's clearer, what do you think ?

**Review by gfyrag — DISMISSED** [review]

File: pkg/storage/utils.go (hunk 16 above)
Comment (antoinegelloz) [reply pkg/storage/utils.go]: It's better thanks :D

File: pkg/storage/transactions.go [hunk 20]
```diff
@@ -0,0 +1,90 @@
+package storage
+
+import (
+	"time"
+)
+
+const (
+	DefaultLimit = 15
```
Comment (antoinegelloz) [inline pkg/storage/transactions.go]:
```suggestion
	QueryDefaultLimit = 15
```

File: pkg/storage/transactions.go [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+type TransactionsQuery struct {
+	Limit     uint
+	AfterTxID uint64
+	Params    map[string]interface{}
+}
+
+type TxModifier func(*TransactionsQuery)
```
Comment (antoinegelloz) [inline pkg/storage/transactions.go]:
```suggestion
type TxQueryModifier func(*TransactionsQuery)
```

File: pkg/storage/accounts.go [hunk 22]
```diff
@@ -0,0 +1,100 @@
+package storage
+
+type AccountsQuery struct {
+	Limit        uint
+	Offset       uint
+	AfterAddress string
+	Params       map[string]interface{}
+}
+
+type AccModifier func(*AccountsQuery)
```
Comment (antoinegelloz) [inline pkg/storage/accounts.go]:
```suggestion
type AccQueryModifier func(*AccountsQuery)
```

File: pkg/api/controllers/account_controller_test.go [hunk 23]
```diff
@@ -168,6 +168,135 @@ func TestGetAccounts(t *testing.T) {
...
+				//	rsp = internal.GetAccounts(api, url.Values{
+				//		"address": []string{"b.b"},
+				//	})
+				//	assert.Equal(t, http.StatusOK, rsp.Result().StatusCode)
+				//	cursor := internal.DecodeCursorResponse[core.Account](t, rsp.Body)
+				//	// 1 accounts: bob
+				//	assert.Len(t, cursor.Data, 1)
+				//	assert.Equal(t, cursor.Data[0].Address, "bob")
+				//})
+
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go]:
```suggestion
```

Comment (jdupas22) [reply pkg/api/controllers/account_controller_test.go]: i'll just uncomment it !

File: pkg/api/controllers/account_controller_test.go [hunk 24]
```diff
@@ -168,6 +168,135 @@ func TestGetAccounts(t *testing.T) {
 					assert.Equal(t, http.StatusBadRequest, rsp.Result().StatusCode, rsp.Body.String())
 				})
 
+				t.Run("address", func(t *testing.T) {
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller_test.go]: You uncommented this test, but it is already there, line 135 😅

File: pkg/storage/accounts.go (hunk 22 above)
Comment (antoinegelloz) [reply pkg/storage/accounts.go]: You haven't changed it haha

File: pkg/api/controllers/account_controller.go [hunk 25]
```diff
@@ -67,16 +68,47 @@ func (ctl *AccountController) GetAccounts(c *gin.Context) {
...
+			if balanceOperator, ok = storage.NewBalanceOperator(balanceOperatorStr); !ok {
+				ResponseError(
+					c,
+					ledger.ValidationError{
+						Msg: "invalid parameter 'balance_operator', should be one of 'e, gt, gte, lt, lte'"})
+				return
+			}
+		} else {
+			// default value for balanceOperator
+			balanceOperator = storage.BalanceOperatorGte
```
Comment (antoinegelloz) [inline pkg/api/controllers/account_controller.go]: Create a `DefaultBalanceOperator`?

**Review by antoinegelloz — DISMISSED** [review]

**Review by antoinegelloz — APPROVED** [review]
